tests/test_query.py

```diff
@@ -0,0 +1,90 @@
+from marshmallow import fields, Schema
+import pytest
+
+from flask_resty import Api, ApiView
+from flask_resty.testing import assert_response
+
+# -----------------------------------------------------------------------------
+
+
+@pytest.fixture
+def schemas():
+    class NameSchema(Schema):
+        name = fields.String(required=True)
+
+    class NameListSchema(Schema):
+        names = fields.List(fields.String(), load_from='name', required=True)
+
+    return {
+        'name': NameSchema(),
+        'name_list': NameListSchema(),
+    }
+
+
+@pytest.fixture(autouse=True)
+def routes(app, schemas):
+    class NameView(ApiView):
+        query_schema = schemas['name']
+
+        def get(self):
+            return self.make_response(self.get_query_data()['name'])
+
+    class NameListView(ApiView):
+        query_schema = schemas['name_list']
+
+        def get(self):
+            return self.make_response(self.get_query_data()['names'])
+
+    api = Api(app)
+    api.add_resource('/name', NameView)
+    api.add_resource('/names', NameListView)
+
+
+# -----------------------------------------------------------------------------
+
+
+def test_get_name_one(client):
+    response = client.get('/name?name=foo')
+    assert_response(response, 200, 'foo')
+
+
+def test_get_name_extra(client):
+    response = client.get('/name?name=foo&ignored=bar')
+    assert_response(response, 200, 'foo')
+
+
+def test_get_names_one(client):
+    response = client.get('/names?name=foo')
+    assert_response(response, 200, ['foo'])
+
+
+def test_get_names_many(client):
+    response = client.get('/names?name=foo&name=bar')
+    assert_response(response, 200, ['foo', 'bar'])
+
+
+# -----------------------------------------------------------------------------
+
+
+def test_error_get_name_missing(client):
+    response = client.get('/name')
+    assert_response(response, 422, [{
+        'code': 'invalid_query',
+        'source': {'parameter': 'name'},
+    }])
+
+
+def test_error_get_name_many(client):
+    response = client.get('/name?name=foo&name=bar')
+    assert_response(response, 422, [{
+        'code': 'invalid_query',
+        'source': {'parameter': 'name'},
+    }])
+
+
+def test_error_get_names_missing(client):
```

Comment:
this is a little bit odd, since the semantic of `required` on a list is never that the list has to be non empty. I think it would make more sense if `required` in this context didn't have any effect. if you want the non empty semantic you can use a custom validator

Reply:
what do you mean? the parameter is missing entirely, right?

Reply:
the user could always set `missing=[]` or something

Reply:
uh, isn't the way of defining lists `foo=1&foo=2` and not `foo=[...]`? in the former case, there is no distinction between "empty" and "missing".

Reply:
well, querystring does, no? you could always do `?foo`

Reply:
that's also ambiguous, because it could be interpreted as `[None]`.

Reply:
well, it's not ideal, but i think i'm ok with matching existing semantics... if you call `args.getlist()` on something missing, you do get an error

Reply:
I guess there isn't an optimal solution. ok with it